File: lclang/src/main/java/lclang/statements/expressions/ContainerExpression.java
package lclang.statements.expressions;

import lclang.Caller;
import lclang.LCBaseExecutor;
import lclang.Link;
import lclang.exceptions.LCLangRuntimeException;
import lclang.libs.lang.classes.VoidClass;
import lclang.statements.Statement;

public class ContainerExpression extends Expression {
    private final Statement[] statements;

    public ContainerExpression(Statement[] statements) {
        super(0);
        this.statements = statements;
    }

    @Override
    public Link visit(Caller prevCaller, LCBaseExecutor visitor) throws LCLangRuntimeException {
        for(Statement stmt: statements) {
            Caller caller = stmt.getCaller(prevCaller);
            Link value = stmt.visit(caller, visitor);
            if(value.state!= Link.State.NOTHING) {
                if(value.state== Link.State.RETURN)
                    value.state = Link.State.NOTHING;

                return value;
            }else value.get.invoke(caller);
        }

        return VoidClass.value;
    }
}

File: lclang/src/main/kotlin/lclang/statements/ContainerExpression.kt
package lclang.statements

import lclang.Caller
import lclang.LCBaseExecutor
import lclang.VOID_VALUE
import lclang.Value

class ContainerExpression(
    val statements: Array<Statement>,
    line: Int,
    column: Int
): Expression(line, column) {
    override fun visit(prevCaller: Caller, visitor: LCBaseExecutor): Value {
        for(stmt in statements) {
            val caller = stmt.getCaller(prevCaller)
            val value = stmt.visit(caller, visitor)
            if(value.state!= Value.State.NOTHING) {
                if(value.state== Value.State.RETURN)
                    value.state = Value.State.NOTHING

                return value
            }else value.get(caller)
        }

        return VOID_VALUE
    }
}